### PR TITLE
Build: Benchmark individual package's sizes and dependencies

### LINKED ISSUE #29322: Set up tooling to warn/highlight on regressions in package size and dependency graph
We need to ensure that new PRs and upcoming work increases our bundle size and dependency graph significantly, now that we've invested so much effort into optimizing and reducing our footprint.

For each PR we want to highlight the following changes, in comparison with the head commit on `next`:

1. Any changes to individual package's dependencies (including transitive), the total dependency install size (if possible) and the total number of dependencies.
2. Any changes to individual package's devDependencies, that could cause the package's own bundle size to increase, because devDependencies are by default bundled in.
3. Include links to the visual metafile analysis story in our UI Storybook, so that bundle size changes can more easily be inspected.

To reduce the noise, we'll only report these when we detect changes in package's `package.json` and/or the global lock-file. The global lock-file can impact the prebundled dependencies, so when we detect changes there, we have to compare and highlight all packages.

The report will be sent as a comment by our GitHub bot

# Example comment

The following packages have changes to their dependencies:

### `@storybook/react`

|  | **Before** | **After** | **Difference** |
|---|---|---|---|
| Bundled size | 273 kB | 300 kB | +28 kB |
| Dependency count | 8 | 12 | +4 |
| Dependency size | 3.5 MB | 3.8 MB | 🚨 **+300 kB** 🚨  |
| Bundle Size Analyzer | "Link to story on `next`" | "Link to story on this branch" |  |

### `@storybook/svelte`

|  | **Before** | **After** | **Difference** |
|---|---|---|---|
| Bundled size | 273 kB | 200 kB | 🎉 -73 kB 🎉  |
| Dependency count | 8 | 12 | +4 |
| Dependency size | 3.5 MB | 1.5 MB | 🎉 -2 MB 🎉  |
| Bundle Size Analyzer | "Link to story on `next`" | "Link to story on this branch" |  |

### PR DESCRIPTION
**Pull request by JReinhold**
Closes #29322

## What I did

Added a `bench-packages` script that will benchmark any or all packages' sizes and dependencies.

```
yarn bench-packages --help
Usage: bench-packages [options] [packages...]

Arguments:
  packages                     which packages to bench. If omitted, all packages are benched

Options:
  -b, --baseBranch <string>    The base branch to compare the results with. Requires GCP_CREDENTIALS env var
  -p, --pull-request <number>  The PR number to add compare results to. Only used together with --baseBranch
  -u, --upload                 Upload results to BigQuery. Requires GCP_CREDENTIALS env var
  -h, --help                   display help for command                                                                                                                                                     
```

Eg. `yarn bench-packages --baseBranch next --pull-request 123 --upload`, or `yarn bench-packages @storybook/core @storybook/react` to only bench a specific set of packages. Uploading results and/or comparing with base branches requires the `GCP_CREDENTIALS` environment variable to be set.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  78.2 MB | 78.2 MB | 60 B | 0.75 | 0% |
| initSize |  143 MB | 143 MB | 46 B | 1.15 | 0% |
| diffSize |  65.2 MB | 65.2 MB | -14 B | **3.03** | 0% |
| buildSize |  6.88 MB | 6.88 MB | -14 B | 0.42 | 0% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | -14 B | -0.42 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.9 MB | 1.9 MB | 0 B | 0.43 | 0% |
| buildSbPreviewSize |  271 kB | 271 kB | 0 B | - | 0% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.88 MB | 3.88 MB | -14 B | 0.42 | 0% |
| buildPreviewSize |  3 MB | 3 MB | 0 B | 0.42 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.3s | 22.7s | 15.3s | **1.31** | 🔺67.6% |
| generateTime |  19.9s | 18.7s | -1s -213ms | -0.93 | -6.5% |
| initTime |  13.5s | 12.8s | -728ms | **-1.28** | 🔰-5.7% |
| buildTime |  8.7s | 8.5s | -201ms | -0.12 | -2.4% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5s | 5.7s | 658ms | -0.07 | 11.5% |
| devManagerResponsive |  3.2s | 3.4s | 256ms | -0.27 | 7.3% |
| devManagerHeaderVisible |  477ms | 588ms | 111ms | 0.45 | 18.9% |
| devManagerIndexVisible |  543ms | 614ms | 71ms | -0.04 | 11.6% |
| devStoryVisibleUncached |  1s | 938ms | -154ms | -0.28 | -16.4% |
| devStoryVisible |  508ms | 610ms | 102ms | 0.04 | 16.7% |
| devAutodocsVisible |  451ms | 486ms | 35ms | -0.26 | 7.2% |
| devMDXVisible |  441ms | 537ms | 96ms | 0.56 | 17.9% |
| buildManagerHeaderVisible |  483ms | 504ms | 21ms | -0.66 | 4.2% |
| buildManagerIndexVisible |  494ms | 517ms | 23ms | -0.66 | 4.4% |
| buildStoryVisible |  484ms | 499ms | 15ms | -0.72 | 3% |
| buildAutodocsVisible |  411ms | 467ms | 56ms | 0.67 | 12% |
| buildMDXVisible |  401ms | 456ms | 55ms | 0.12 | 12.1% |

## Greptile Summary

Here's my concise review of the PR that adds package benchmarking functionality:

Adds a new benchmarking system to measure and compare package sizes and dependencies across branches, with support for uploading results to BigQuery and GitHub for analysis.

- Added `scripts/bench/bench-packages.ts` implementing core benchmarking logic with size/dependency measurements and comparison thresholds
- Added new `bench-packages` CircleCI job in `.circleci/config.yml` that runs after build step
- Updated `scripts/run-registry.ts` to use async file cleanup with `fs/promises` instead of sync operations
- Added `bench-packages` script and `tinyexec` dependency in `scripts/package.json` for command execution

The implementation looks solid with good error handling and configurability, though there may be some edge cases around async cleanup that could be improved.

### CODE REVIEW

File: scripts/run-registry.ts [hunk 1]
```diff
@@ -207,7 +207,7 @@ const run = async () => {
     await publish(packages, 'http://localhost:6002');
   }
 
-  await execa('npx', ['rimraf', '.npmrc'], { cwd: root });
+  await rm(join(root, '.npmrc'), { force: true });
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply scripts/run-registry.ts:210]: This makes a lot more sense indeed, @JReinhold

File: scripts/run-registry.ts [hunk 2]
```diff
@@ -217,6 +217,7 @@ const run = async () => {
 
 run().catch((e) => {
   logger.error(e);
-  execaSync('npx', ['rimraf', '.npmrc'], { cwd: root });
-  process.exit(1);
+  rm(join(root, '.npmrc'), { force: true }).then(() => {
+    process.exit(1);
+  });
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply scripts/run-registry.ts:222]: Probably just needs an `await`

[the remark this answers is not in the document]
Comment (JReinhold) [reply scripts/run-registry.ts:222]: top-level await isn't supported in our env. I think this comment is nonsense.

File: scripts/run-registry.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (JReinhold) [reply scripts/run-registry.ts:210]: this doesn't make any sense, `run()` already has plenty of other async operations, it wouldn't work at all if async stuff weren't awaited in here.

[the remark this answers is not in the document]
Comment (ndelangen) [reply scripts/run-registry.ts:210]: Yeah, the new code is a lot better. sorry for the confusion

**Review by ndelangen — APPROVED** [review]

**Pull request merged**